Anonymization Improvements

Ashod opened the pull request:
- wsd: move anonymization data to Anonymizer.cpp
- wsd: privatize the anonimization globals
- wsd: fuzzer: fix dependency on Anonymizer
- wsd: enable anonymization with trace and config flag
- wsd: initialize the anonymizer
- wsd: logging
- wsd: move the anonymizer prefix counter to the class
- wsd: remove unused salt argument
- wsd: test: reinitialize the anonymizer
- wsd: remove unnecessary anonymization bits

caolanm commented:
android and iOS makefile equivalents presumably need Anonymizer.cpp added into them

vmiklos reviewed (approved):
1) As Caolan points out, seems you only updated the automake build system for the new .cpp files, while Android uses cmake and iOS uses xcode, so that needs to be tweaked similarly.

2) Unless you work on a security fix, it's useful to give a "why" (reason) for your changes in the commit message: the first line is just a summary. If you leave the body of your commit message empty, the next person (or your future self) can be frustrated, please don't do that. :-) E.g. the "wsd: test: reinitialize the anonymizer" commit doesn't say why it does so.

3) Other than these, see inline comments for nits, but once this passes CI, I see nothing obviously problematic with this.

Ashod commented on wsd/COOLWSD.cpp:
```diff
@@ -1439,11 +1439,8 @@ void COOLWSD::innerInitialize(Poco::Util::Application& self)
-                    "This will leak user-data!");
-
-            // Disable anonymization as it's useless now.
-            AnonymizeUserData = false;
```
Fair enough. I added description.

(This whole PR is moving the Anonymization globals into the class, so this was natural to delete. But you are right, it might not be obvious when the patch is viewed in isolation.)

Ashod commented on common/Anonymizer.hpp:
```diff
@@ -156,6 +160,9 @@ class Anonymizer
     static std::unique_ptr<Anonymizer> _instance;
 
+    /// The prefix counter.
+    std::atomic<unsigned> _prefix;
```
This minimizes the scope of the lock. That lock can have a significant performance impact, so making this an atomic is a good idea. Otherwise, we have to take the lock over a much larger code (and for much longer time), or take it multiple times, locking and unlocking, which again is highly costly and problematic.

Ashod commented:
> 1. As Caolan points out, seems you only updated the automake build system for the new .cpp files, while Android uses cmake and iOS uses xcode, so that needs to be tweaked similarly.
> 2. Unless you work on a security fix, it's useful to give a "why" (reason) for your changes in the commit message: the first line is just a summary. If you leave the body of your commit message empty, the next person (or your future self) can be frustrated, please don't do that. :-) E.g. the "wsd: test: reinitialize the anonymizer" commit doesn't say why it does so.
> 3. Other than these, see inline comments for nits, but once this passes CI, I see nothing obviously problematic with this.

Thanks for the feedback, @vmiklos. Addressed your points.

timar commented:
I don't know what to think about this persistent CI failure. 
- I restarted Jenkins
- I restarted the build nodes
- I rebased to latest master
On the other hand, on the same build node [Build 3984](https://cpci.cbg.collabora.co.uk:8080/job/github_online_master_debug_vs_co-24.04/3984/) was successful for PR #10766 and the subsequent [Build 3985](https://cpci.cbg.collabora.co.uk:8080/job/github_online_master_debug_vs_co-24.04/3985/) was unsuccessful for this PR. It is not obvious to me that the problem is unrelated to this PR.

vmiklos commented:
https://cpci.cbg.collabora.co.uk:8080/job/github_online_master_debug_vs_co-24.04/3988/console fails at:

> FATAL: command execution failed

Is it possible that we sent some signal to the wrong process and that ends up being the parent of the test, so the Jenkins runner gets killed?

'git grep forwardSig' has a number of hits, I wonder if we could somehow log what signals we send to what processes, what's the name of those processes, and perhaps that would help.

Or maybe easiest if @Ashod does an ssh to the CI machine if the problem is visible there, so it can be debugged interactively?

And sure, possibly this PR just exposes an already existing problem. I hope one of these helps.

timar commented:
Running `make check -j 1`:
```
Running unit-wopi-async-slow | unit-wopi-async-slow.log | unit-wopi-async-slow.trs ../test/run_unit.sh --test-name unit-wopi-async-slow.la --log-file unit-wopi-async-slow.log --trs-file unit-wopi-async-slow.trs --color-tests yes --enable-hard-errors yes --expect-failure no -- ./unit-wopi-async-slow.la
make: *** [check-recursive] User defined signal 1
make[2]: *** [check-TESTS] User defined signal 1
make[3]: *** [unit-wopi-async-slow.log] User defined signal 1
User defined signal 1
```
the parent make process gets the User defined signal 1! Actually it's a cascade of signals, because in the end even the login process is killed.

Ashod commented:
Yes, it seems that `kill(pid, 0)` is causing pain. No idea why. But you are both correct. It's reaching the login and killing it. Very odd.

Ashod commented:
All green, now that the offending patch is removed. Thanks for the investigation everyone!
